### ISSUE TITLE
DataViewCustomRenderer of varianttype PyObject and empty lists

### ISSUE THREAD

**Issue by jensgoe** [wxWidgets/Phoenix#1599]
**Operating system**: Win10
**wxPython version & source**:          4.1.0a1 (master)
**Python version & source**:            3.7.4

**Description of the problem**:

I have DataViewCtrl with one column using a CustomRenderer and varianttype PyObject. My Model is returning tuples/list containing Python-objects. In the sample below just numbers. GetValue() of MyModel returns an empty tuple for the first row. For the second row a tuple with one number (666,), Third row a tuple with 2 numbers (666, 666). Running the sample results in tracebacks during rendering "TypeError: object of type 'NoneType' has no len()"

I did some debugging and found out that i_wxVariant_in_helper checks for several types to convert a python object to a wxVariant. The check at wxPyCheckStringSequence now gets an empty list and this seems ok to be a string sequence and so a wxVariant of type arrstring is returned. Then wxWidgets checks if the wxVariant coming from the model fits to the varianttype of the renderer. Thats not the case and the renderer receives None in its SetValue method. 

I am not sure whats a good fix. Modifying wxPyCheckStringSequence to handle empty sequences not as sequences would probably my favorite. But as soon Model.GetValue() returns really a list of string a Variant of type arrstring is created and i have a conflict to PyObject again. Maybe removing StringSequence (respectively varianttype arrstring) completely, because it is a PyObject in Python. But i am not sure if this results in problems with wxDataViewChoiceRenderer that expects a wxArrayString)

```python

import wx
import wx.dataview as dv


class MyCustomRenderer(dv.DataViewCustomRenderer):
    def __init__(self):
        dv.DataViewCustomRenderer.__init__(self, "PyObject")
        self._value = ()
    def SetValue(self, value):
        self._value = value
        return True
    def GetSize(self):
        return wx.Size(100 * len(self._value), 20)
    def Render(self, cell, dc, state):
        self.RenderText(str(self._value), 0, cell, dc, state)
        return True


class MyModel(dv.PyDataViewModel):
    def GetChildren(self, item, children):
        if item == dv.NullDataViewItem:
            children.append(self.ObjectToItem(0))
            children.append(self.ObjectToItem(1))
            children.append(self.ObjectToItem(2))
            return 3
        return 0
    def GetValue(self, item, col):
        obj = self.ItemToObject(item)
        result = (666,) * obj
        return result
    def GetColumnCount(self):
        return 1
    def IsContainer(self, item):
        if item == dv.NullDataViewItem:
            return True
        else:
            return False


if __name__ == "__main__":
    app = wx.App()

    frame = wx.Frame(None)
    dvc = dv.DataViewCtrl(frame)
    model = MyModel()
    dvc.AssociateModel(model)
    
    renderer = MyCustomRenderer()
    col = dv.DataViewColumn('Col1', renderer, 0, width=320)
    dvc.AppendColumn(col)
    
    frame.Show(True)

    app.SetTopWindow(frame)
    app.MainLoop()
```

**Comment by jensgoe** [comment]
i have prepared an unittest in commit 7f1beab024b5acec44aee51f8f98943183eacf7a

**Comment by RobinD42** [comment]
Yeah, the order that types are tried when converting an object to a `wxVariant` tends to be a little fragile. When I fix one case it tends to cause a different problem in another case. We also have to keep in mind the needs of the propgrid classes, which also use `wxVariant`.

There's a change coming soon (maybe today) that could help out in this case. See #1600. It allows overriding `HasValue` so `GetValue` will not even be called for cells that are empty.

Another way to force the use of a `PyObject` variant would be to embed your sequence values in another object. For example, your model's `GetValue` could do something like this `return MyObject(result)` and then in the renderer just grab the sequence value from an attribute of the `MyObject` that is passed in.

I'll take another look at the `wxVariant` conversion code, but I'm a little reluctant to change things there this close to a release.

**Comment by jmoraleda** [comment]
For what is worth, a different example of what I think is the same issue at the root:  I have a python `datetime.datetime` that I don't want converted to a `wx.DateTime` (because I lose timezone information) but specifying `PyObject` (vs `datetime`) as the variant type does not work because it still gets converted and I just get `None`.
